text/0027-corejam.md

```diff
@@ -0,0 +1,706 @@
+
+The Join-Accumulate stage may be seen as a synchronized counterpart to the parallelised Collect-Refine stage. It may be used to integrate the work done from the context of an isolated VM into a self-consistent singleton world model. In concrete terms this means ensuring that the independent work components, which cannot have been aware of each other during the Collect-Refine stage, do not conflict in some way. Less dramatically, this stage may be used to enforce ordering or provide a synchronisation point (e.g. for combining entropy in a sharded RNG). Finally, this stage may be a sensible place to manage asynchronous interactions between subcomponents of a Work Class or even different Work Classes and oversee message queue transitions.
+
+#### Initial Validation
+
+There are a number of initial validation requirements which the RCBA must do in order to ensure no time is wasted on further, possibly costly, computation.
+
+Firstly, any given Work Report must have enough attestation signatures to be considered for inclusion on-chain. Only one Work Report may be considered for inclusion from each RCVG per block.
+
+Secondly, any Work Reports introduced by the RCBA must be *Recent*, defined as having a `context.header_hash` which is an ancestor of the RCBA head and whose height is less than `RECENT_BLOCKS` from the block which the RCBA is now authoring.
+
+```rust
+const RECENT_BLOCKS: u32 = 16;
+```
+
+Thirdly, the RCBA may not include multiple Work Reports for the same Work Package. Since Work Reports become inherently invalid once they are no longer *Recent*, then this check may be simplified to ensuring that there are no Work Reports of the same Work Package within any *Recent* blocks.
+
+Fourthly, the RCBA may not include Work Reports whose prerequisites are not themselves included in *Recent* blocks.
+
+In order to ensure all of the above tests are honoured by the RCBA, a block which contains Work Reports which fail any of these tests shall panic on import. The Relay-chain's on-chain logic will thus include these checks in order to ensure that they are honoured by the RCBA. We therefore introduce the *Recent Inclusions* storage item, which retaining all Work Package hashes which were included in the *Recent* blocks:
+
+```rust
+const MAX_CORES: u32 = 512;
+/// Must be ordered.
+type InclusionSet = BoundedVec<WorkPackageHash, ConstU32<MAX_CORES>>;
+type RecentInclusions = StorageValue<BoundedVec<InclusionSet, ConstU32<RECENT_BLOCKS>>>
+```
+
+The RCBA must keep an up to date set of which Work Packages have already been included in order to avoid accidentally attempting to introduce a duplicate Work Package or one whose prerequisites have not been fulfilled. Since the currently authored block is considered *Recent*, Work Reports introduced earlier in the same block do satisfy prerequisites of Work Packages introduced later.
```

Comment:
It seems that pre-requisites could be fulfilled both by earlier and later Work Reports within the same block, and this introduces only a small linear overhead. Algorithm sketch:
  1. Keep a set of all unfulfilled pre-requisites for all Work Reports being posted to the chain
  2. When a work report's hash is present in the set, remove that from the set
  3. If the set is non-empty at the end of the block, panic.

Reply:
(pre-requisite handling is hard - also discussed in detail in DM. we might better punt this off to another RFC)